core/lib/zksync_core/src/api_server/web3/mod.rs

```diff
@@ -730,24 +510,37 @@ impl<G: 'static + Send + Sync + L1GasPriceProvider> ApiBuilder<G> {
             .layer(in_flight_requests)
             .option_layer(cors);
 
-        let server_builder = if is_http {
-            ServerBuilder::default().http_only().max_connections(5_000)
+        let (local_addr, server_handle) = if is_http {
+            let server_builder = ServerBuilder::default().http_only().max_connections(5_000);
+            let server = server_builder
+                .set_batch_request_config(batch_request_config)
+                .set_http_middleware(middleware)
+                .set_id_provider(EthSubscriptionIdProvider::default())
+                .max_response_body_size(response_body_size_limit)
+                .build(addr)
+                .await
+                .with_context(|| format!("Failed building {transport_str} JSON-RPC server"))?;
+            (server.local_addr(), server.start(rpc))
         } else {
-            ServerBuilder::default().ws_only()
+            let server_builder = ServerBuilder::default().ws_only().set_rpc_middleware(
+                RpcServiceBuilder::new().layer_fn(move |a| {
+                    LimitMiddleware::new(a, websocket_requests_per_minute_limit)
+                }),
+            );
+            let server = server_builder
+                .set_batch_request_config(batch_request_config)
+                .set_http_middleware(middleware)
+                .set_id_provider(EthSubscriptionIdProvider::default())
+                .max_response_body_size(response_body_size_limit)
+                .build(addr)
+                .await
+                .with_context(|| format!("Failed building {transport_str} JSON-RPC server"))?;
+            (server.local_addr(), server.start(rpc))
```

Comment:
This `if` `else` is neeeded because otherwise inserting the `LimitMiddleware` will result in `ServerBuilder` types mismatch. So we need to go all the way to actually creating the server to avoid that.